tests/components/modbus/test_init.py

```diff
@@ -480,113 +480,97 @@ async def test_pymodbus_connect_fail(hass, caplog, mock_pymodbus):
-    now = now + timedelta(seconds=6)
+        assert hass.states.get(entity_id).state == STATE_UNAVAILABLE
+
+    stop_time = start_time + timedelta(seconds=test_delay)
+    step_timedelta = timedelta(seconds=1)
+    while now < stop_time:
+        now = now + step_timedelta
+        with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
+            async_fire_time_changed(hass, now)
+            await hass.async_block_till_done()
+            assert hass.states.get(entity_id).state == STATE_UNAVAILABLE
+    now = now + step_timedelta + step_timedelta
     with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
         async_fire_time_changed(hass, now)
         await hass.async_block_till_done()
-    now = now + timedelta(seconds=10)
+        assert hass.states.get(entity_id).state == STATE_ON
+
+
+async def test_thread_lock(hass, mock_pymodbus):
+    """Run test for block of threads."""
+
+    # the purpose of this test is to test the threads are not being blocked
+    # We "hijiack" a binary_sensor to make a proper blackbox test.
+    test_scan_interval = 5
+    sensors = []
+    for i in range(200):
+        sensors.append(
+            {
+                CONF_INPUT_TYPE: CALL_TYPE_COIL,
+                CONF_NAME: f"{TEST_SENSOR_NAME}_{i}",
+                CONF_ADDRESS: 52 + i,
+                CONF_SCAN_INTERVAL: test_scan_interval,
+            }
+        )
+    config = {
+        DOMAIN: [
+            {
+                CONF_TYPE: "tcp",
+                CONF_HOST: "modbusTestHost",
+                CONF_PORT: 5501,
+                CONF_NAME: TEST_MODBUS_NAME,
+                CONF_BINARY_SENSORS: sensors,
+            }
+        ]
+    }
+    mock_pymodbus.read_coils.return_value = ReadResult([0x01])
+    now = dt_util.utcnow()
     with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
-        async_fire_time_changed(hass, now)
+        assert await async_setup_component(hass, DOMAIN, config) is True
         await hass.async_block_till_done()
-
-    # Check states
-    entity_id = f"{BINARY_SENSOR_DOMAIN}.{TEST_SENSOR_NAME}_1"
-    assert not hass.states.get(entity_id).state == STATE_UNAVAILABLE
-    entity_id = f"{BINARY_SENSOR_DOMAIN}.{TEST_SENSOR_NAME}_2"
-    assert not hass.states.get(entity_id).state == STATE_UNAVAILABLE
-    entity_id = f"{SENSOR_DOMAIN}.{TEST_SENSOR_NAME}_3"
-    assert not hass.states.get(entity_id).state == STATE_UNAVAILABLE
-    entity_id = f"{SENSOR_DOMAIN}.{TEST_SENSOR_NAME}_4"
-    assert not hass.states.get(entity_id).state == STATE_UNAVAILABLE
+    stop_time = now + timedelta(seconds=10)
+    step_timedelta = timedelta(seconds=1)
+    while now < stop_time:
+        now = now + step_timedelta
+        with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
+            async_fire_time_changed(hass, now)
+            await hass.async_block_till_done()
```

Comment:
Maybe assert that the sensor states have the correct state at the end of the test?

Reply:
Good idea, done